src/Stratis.Bitcoin.Features.Consensus/Rules/ConsensusRule.cs

```diff
@@ -0,0 +1,59 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Features.Consensus.Rules.CommonRules;
+
+namespace Stratis.Bitcoin.Features.Consensus.Rules
+{
+    /// <summary>
+    /// An abstract rule for allowing to write consensus rules.
+    /// </summary>
+    public abstract class ConsensusRule
+    {
+        /// <summary>Instance logger.</summary>
+        public ILogger Logger { get; set; }
+
+        /// <summary>The engine this rule belongs to.</summary>
+        public ConsensusRules Parent { get; set; }
+
+        /// <summary>
+        /// A collection of rules that the current rule depends on.
+        /// </summary>
+        /// <returns>A collection of rules this rule depends on.</returns>
+        public virtual IEnumerable<Type> Dependencies()
```

Comment:
Will that mean `the rules in this list need to be defined to be run at some point` or `the rules in this list need to have been run already`?

Reply:
Ah yes good point, to me it means the Rules have to exist and have run before this rule.
Some rules may update the context with stuff and later rules may depend on that updated context.
Of also rules that are cheep to execute we want to check first.

I considered using such a construction `ConsensusRule<TDependentRule>` where  TDependentRule is the dependent rule, how ever this wont work in case of more then one rule dependency.
We could also provide a few of such base classes `ConsensusRule<TDependentRule1,TDependentRule2`>